extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java

```diff
@@ -0,0 +1,1263 @@
+    Map<Integer, Long> partitionOffsets = new HashMap<>();
+
+    Map<String, TaskData> tasks = new HashMap<>();
+    DateTime completionTimeout; // is set after signalTasksToFinish(); if not done by timeout, take corrective action
+
+    public TaskGroup(Map<Integer, Long> partitionOffsets)
+    {
+      this.partitionOffsets = partitionOffsets;
+    }
+  }
+
+  private class TaskData
+  {
+    TaskStatus status;
+    DateTime startTime;
+  }
+
+  // Map<{group ID}, {actively reading task group}>; see documentation for TaskGroup class
+  private final HashMap<Integer, TaskGroup> taskGroups = new HashMap<>();
+
+  // After telling a taskGroup to stop reading and begin publishing a segment, it is moved from [taskGroups] to here so
+  // we can monitor its status while we queue new tasks to read the next range of offsets. This is a list since we could
+  // have multiple sets of tasks publishing at once if time-to-publish > taskDuration.
+  // Map<{group ID}, List<{pending completion task groups}>>
+  private final HashMap<Integer, List<TaskGroup>> pendingCompletionTaskGroups = new HashMap<>();
+
+  // The starting offset for a new partition in [partitionGroups] is initially set to null. When a new task group
+  // is created and is assigned partitions, if the offset in [partitionGroups] is null it will take the starting
+  // offset value from the metadata store, and if it can't find it there, from Kafka. Once a task begins
+  // publishing, the offset in partitionGroups will be updated to the ending offset of the publishing-but-not-yet-
+  // completed task, which will cause the next set of tasks to begin reading from where the previous task left
+  // off. If that previous task now fails, we will set the offset in [partitionGroups] back to null which will
+  // cause successive tasks to again grab their starting offset from metadata store. This mechanism allows us to
+  // start up successive tasks without waiting for the previous tasks to succeed and still be able to handle task
+  // failures during publishing.
+  // Map<{group ID}, Map<{partition ID}, {startingOffset}>>
+  private Map<Integer, Map<Integer, Long>> partitionGroups = new HashMap<>();
+  // --------------------------------------------------------
+
+  private final TaskStorage taskStorage;
+  private final TaskMaster taskMaster;
+  private final IndexerMetadataStorageCoordinator indexerMetadataStorageCoordinator;
+  private final KafkaIndexTaskClient taskClient;
+  private final ObjectMapper sortingMapper;
+  private final KafkaSupervisorSpec spec;
+  private final String dataSource;
+  private final KafkaSupervisorIOConfig ioConfig;
+  private final String supervisorId;
+
+  private final ExecutorService exec;
+  private final BlockingQueue<Notice> notices = new LinkedBlockingDeque<>();
+  private final Timer timer = new Timer(true);
+  private final Object stopLock = new Object();
+
+  private KafkaConsumer consumer;
```

Comment:
This should be volatile since it's initialized in the thread calling `start` and then used in a different thread after that.